jf_agent/git/github_gql_adapter.py

```diff
@@ -0,0 +1,452 @@
+        users = [
+            _normalize_user(user)
+            for project_id in self.config.git_include_projects
+            for user in self.client.get_users(project_id)
+        ]
+        print('✓')
+        return users
+
+    @diagnostics.capture_timing()
+    @agent_logging.log_entry_exit(logger)
+    def get_repos(
+        self, normalized_projects: List[NormalizedProject],
+    ) -> List[NormalizedRepository]:
+        print('downloading github repos... ', end='', flush=True)
+
+        nrm_repos: List[NormalizedRepository] = []
+
+        filters = []
+        if self.config.git_include_repos:
+            filters.append(
+                lambda r: r['name'].lower()
+                in set([r.lower() for r in self.config.git_include_repos])
+            )
+        if self.config.git_exclude_repos:
+            filters.append(
+                lambda r: r['name'].lower()
+                not in set([r.lower() for r in self.config.git_exclude_repos])
+            )
+
+        for nrm_project in normalized_projects:
+            for api_repo in tqdm(
+                self.client.get_repos(nrm_project.login, repo_filters=filters),
+                desc=f'downloading repos for {nrm_project.login}',
+                unit='repos',
+            ):
+                # Enter the repo to the ID to Name look up, incase the repo name gets
+                # scrubbed by our git_redact_names_and_urls logic
+                self.repo_id_to_name_lookup[api_repo['id']] = api_repo['name']
+                nrm_repos.append(
+                    _normalize_repo(api_repo, nrm_project, self.config.git_redact_names_and_urls)
+                )
+
+        print('✓')
+        if not nrm_repos:
+            raise ValueError(
+                'No repos found. Make sure your token has appropriate access to Github and check your configuration of repos to pull.'
+            )
+        return nrm_repos
+
+    @diagnostics.capture_timing()
+    @agent_logging.log_entry_exit(logger)
+    def get_commits_for_included_branches(
+        self,
+        normalized_repos: List[NormalizedRepository],
+        included_branches: dict,
+        server_git_instance_info,
+    ) -> List[NormalizedCommit]:
+        print('downloading github commits on included branches... ', end='', flush=True)
+        for i, nrm_repo in enumerate(normalized_repos, start=1):
+            with agent_logging.log_loop_iters(logger, 'repo for branch commits', i, 1):
+                pull_since = pull_since_date_for_repo(
+                    server_git_instance_info, nrm_repo.project.login, nrm_repo.id, 'commits'
+                )
+
+                try:
+                    for branch_name in get_branches_for_normalized_repo(
+                        nrm_repo, included_branches
+                    ):
+                        for j, api_commit in enumerate(
+                            tqdm(
+                                self.client.get_commits(
```

Comment:
The control flow here will fail to get the entire branch if we error out for any commit on the branch. Is that our intended behavior? It may be more robust to do

```
for branches
    try get commits
    for commits
        ...

```

Reply:
Oh good call, I'll rework that

Reply:
I reworked it to match how it was functioning before under `jf_agent/git/github.py`. Now if we run into any issues we will still potentially will lose commits but we won't exist early on processing branches. Here is the snippet from the 'legacy' way. The get commits call returns a generator, so it's a little awkward to wrap it in a standalone try-catch block like I think you suggested. I also don't want to put it in the client itself, because I want users of the client to know when we fail to get a commit

```
            for branch in branches_to_process:
                try:
                    for j, commit in enumerate(
                        tqdm(
                            client.get_commits(
                                repo['full_name'], branch, since=pull_since, until=None
                            ),
                            desc=f'downloading commits on branch {branch} for {repo["name"]}',
                            unit='commits',
                        ),
                        start=1,
                    ):
                    ...
```

Reply:
Got it - I think preserving old behavior is the right call here to avoid changing too many things at once. In the future we probably will want to separately make a call on whether the overall control flow, and when we choose to try/except vs. raise, is optimal. My suspicion is that it could be a lot better.